File: leet-code/241. Different Ways to Add Parentheses/DifferentWaysToAddParentheses241_2.java
package com.example;

import java.util.*;

public class DifferentWaysToAddParentheses241_2 {

    private String[] numbers;

    private String[] operators;

    private final Map<String, List<Integer>> memo = new HashMap<>();

    public List<Integer> diffWaysToCompute(String expression) {
        numbers = expression.split("[-+*]");
        operators = Arrays
                .stream(expression.split("\\d+"))
                .filter(s -> !s.isEmpty()) // [, -, -] 에서 앞의 , 를 제거하기 위함
                .toArray(String[]::new);
        return recursion(0, numbers.length - 1);
    }

    private List<Integer> recursion(int low, int high) {
        if (low == high) return List.of(Integer.parseInt(numbers[low]));

        String memoKey = Arrays.stream(numbers, low, high).toString();
        if (memo.containsKey(memoKey)) return memo.get(memoKey);

        List<Integer> res = new ArrayList<>();
        for (int mid = low; mid < high; mid++) {
            List<Integer> left = recursion(low, mid);
            List<Integer> right = recursion(mid + 1, high);

            for (Integer l : left) {
                for (Integer r : right) {
                    if (operators[mid].equals("+")) {
                        res.add(l + r);
                    } else if (operators[mid].equals("-")) {
                        res.add(l - r);
                    } else {
                        res.add(l * r);
                    }
                }
            }
        }
        memo.put(memoKey, res);
        return res;
    }
}

File: leet-code/241. Different Ways to Add Parentheses/DifferentWaysToAddParentheses241_2.kt
package com.example

class DifferentWaysToAddParentheses241_2 {
    private val memo = mutableMapOf<String, List<Int>>()

    fun diffWaysToCompute(expression: String): List<Int> {
        if (memo.containsKey(expression)) {
            return memo[expression]!!
        }
        val result = mutableListOf<Int>()
        for (i in expression.indices) {
            if (expression[i] in "+-*") {
                val left = diffWaysToCompute(expression.substring(0, i))
                val right = diffWaysToCompute(expression.substring(i + 1, expression.length))
                for (l in left) {
                    for (r in right) {
                        when (expression[i]) {
                            '+' -> result.add(l + r)
                            '-' -> result.add(l - r)
                            '*' -> result.add(l * r)
                        }
                    }
                }
            }
        }
        if (result.isEmpty()) {
            result.add(expression.toInt())
        }
        memo[expression] = result
        return result
    }
}
